libs/template/renderer.go

```diff
@@ -0,0 +1,260 @@
+
+	// Execute template and get result
+	result := strings.Builder{}
+	err = tmpl.Execute(&result, r.config)
+	if err != nil {
+		return "", err
+	}
+	return result.String(), nil
+}
+
+func (r *renderer) computeFile(relPathTemplate string) (*inMemoryFile, error) {
+	// read template file contents
+	templateReader, err := r.templateFiler.Read(r.ctx, relPathTemplate)
+	if err != nil {
+		return nil, err
+	}
+	contentTemplate, err := io.ReadAll(templateReader)
+	if err != nil {
+		return nil, err
+	}
+
+	// execute the contents of the file as a template
+	content, err := r.executeTemplate(string(contentTemplate))
+	// Capture errors caused by the "fail" helper function
+	if target := (&ErrFail{}); errors.As(err, target) {
+		return nil, target
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to compute file content for %s. %w", relPathTemplate, err)
+	}
+
+	// Execute relative path template to get materialized path for the file
+	relPath, err := r.executeTemplate(relPathTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	return &inMemoryFile{
+		path:    relPath,
+		content: []byte(content),
+	}, nil
+}
+
+func (r *renderer) walk() error {
+	directories := []string{"."}
+	var currentDirectory string
+
+	for len(directories) > 0 {
+		currentDirectory, directories = directories[0], directories[1:]
+
+		// Skip current directory if it matches any of accumulated skip patterns
+		instanceDirectory, err := r.executeTemplate(currentDirectory)
+		if err != nil {
+			return err
+		}
+		isSkipped, err := r.isSkipped(instanceDirectory)
+		if err != nil {
+			return err
+		}
+		if isSkipped {
+			logger.Infof(r.ctx, "skipping walking directory: %s", instanceDirectory)
+			continue
+		}
+
+		// Add skip function, which accumulates skip patterns relative to current
+		// directory
+		r.baseTemplate.Funcs(template.FuncMap{
+			"skip": func(relPattern string) string {
+				// patterns are specified relative to current directory of the file
+				// {{skip}} function is called from
+				pattern := path.Join(currentDirectory, relPattern)
+				if !slices.Contains(r.skipPatterns, pattern) {
+					logger.Infof(r.ctx, "adding skip pattern: %s", pattern)
+					r.skipPatterns = append(r.skipPatterns, pattern)
+				}
+				// return empty string will print nothing at function call site
+				// when executing the template
+				return ""
+			},
+		})
+
+		// Process all entries in current directory
+		//
+		// 1. For files: the templates in the file name and content are executed, and
+		//     a in memory representation of the file is generated
+		//
+		// 2. For directories: They are appended to a slice, which acts as a queue
+		//     allowing BFS traversal of the template file tree
+		entries, err := r.templateFiler.ReadDir(r.ctx, currentDirectory)
+		if err != nil {
+			return err
+		}
+		for _, entry := range entries {
+			if entry.IsDir() {
+				// Add to slice, for BFS traversal
+				directories = append(directories, path.Join(currentDirectory, entry.Name()))
```

Comment:
We can use `path.Join` because this uses the filer right?

Reply:
Yup, the filer has `filepath.FromSlash` in it to translate windows paths